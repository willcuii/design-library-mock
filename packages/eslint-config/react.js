const { resolve } = require("node:path");

const {
  INTERNAL_RULES,
  IMPORT_RULES,
  PERFECTIONIST_RULES,
  SIMPLE_IMPORT_SORT_RULES,
  UNUSED_IMPORTS_RULES,
} = require("./constants");

const project = resolve(process.cwd(), "tsconfig.json");

/*
 * This is a custom ESLint configuration for use a library
 * that utilizes React.
 */

/** @type {import("eslint").Linter.Config} */
module.exports = {
  extends: [
    "airbnb",
    "airbnb-typescript",
    "prettier",
    "plugin:import/recommended",
    "plugin:import/typescript",
    "plugin:react/recommended",
    "plugin:react-hooks/recommended",
  ],
  parserOptions: {
    project,
  },
  plugins: [
    "only-warn",
    "@stylistic",
    "import",
    "unused-imports",
    "perfectionist",
    "react",
    "react-refresh",
    "simple-import-sort",
  ],
  globals: {
    React: true,
    JSX: true,
  },
  settings: {
    "import/resolver": {
      typescript: {
        project,
      },
    },
  },
  ignorePatterns: ["node_modules/", "dist/", ".eslintrc.js", "**/*.css"],
  // add rules configurations here
  rules: {
    ...INTERNAL_RULES,
    ...IMPORT_RULES,
    ...PERFECTIONIST_RULES,
    ...SIMPLE_IMPORT_SORT_RULES,
    ...UNUSED_IMPORTS_RULES,
    "react-refresh/only-export-components": "warn",
    "react/react-in-jsx-scope": "off",
  },
};
