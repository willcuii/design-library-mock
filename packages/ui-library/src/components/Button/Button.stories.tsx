import React from 'react';
import type {Meta, StoryObj} from '@storybook/react';
import styled from '@emotion/styled';

import Button from './Button';
import {ButtonVariant} from './Button.type';
import {capitalize} from '../../utils/stringUtils';

const meta: Meta<typeof Button> = {
  component: Button,
};

export default meta;
type Story = StoryObj<typeof Button>;

const Row = styled.div`
  display: flex;
  align-items: flex-start;
  flex-wrap: wrap;
  max-width: 600px;
`;

const Column = styled.div`
  flex: 1;
  display: flex;
  flex-direction: column;
  margin: 20px 100px 20px 0;
`;

const Title = styled.h4`
  margin-bottom: 10px;
`;

export const Overview = () => {
  const typeMap = ['gradient', 'primary', 'outlined', 'text'];
  const sizeMap = ['large', 'medium', 'small', 'xsmall'];

  const handleClick = () => {
    console.log('Hello Prenuvo');
  };

  const renderButton = (type: ButtonVariant) => (
    <Column>
      <h2>{capitalize(type)}</h2>
      {sizeMap.map((size) => (
        <Row>
          <Column>
            <Title>Default</Title>
            <Button btnVariant={type} btnSize={size} onClick={handleClick}>
              Button
            </Button>
          </Column>
          <Column>
            <Title>Disabled</Title>
            <Button
              btnVariant={type}
              btnSize={size}
              onClick={handleClick}
              disabled
            >
              Button
            </Button>
          </Column>
        </Row>
      ))}
    </Column>
  );

  return (
    <div style={{padding: '20px 50px'}}>
      <h1>Button Overview</h1>
      <br />
      <li>Can contain an optional leading icon</li>
      <li>Three types: gradient, outlined, and text</li>
      <li>Keep labels concise and in sentence-case</li>
      <li>
        Containers have fully rounded corners and are wide enough to fit label
        text
      </li>
      {typeMap.map((type) => renderButton(type))}
    </div>
  );
};

export const Primary: Story = {
  args: {
    onClick: () => {},
    children: 'Button',
  },
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/fNjLK6NuE2xXm2MsQQkBVF/Halo-UI-1.0.0?type=design&node-id=876-14110&mode=design&t=zYpAhBY2yjbA57Sl-0',
    },
  },
};

export const Outlined: Story = {
  args: {
    onClick: () => {},
    children: 'Button',
    btnVariant: 'outlined',
  },
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/fNjLK6NuE2xXm2MsQQkBVF/Halo-UI-1.0.0?type=design&node-id=876-14110&mode=design&t=zYpAhBY2yjbA57Sl-0',
    },
  },
};

export const Text: Story = {
  args: {
    onClick: () => {},
    children: 'Button',
    btnVariant: 'text',
  },
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/fNjLK6NuE2xXm2MsQQkBVF/Halo-UI-1.0.0?type=design&node-id=876-14110&mode=design&t=zYpAhBY2yjbA57Sl-0',
    },
  },
};

export const Gradient: Story = {
  args: {
    onClick: () => {},
    children: 'Button',
    btnVariant: 'gradient',
  },
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/fNjLK6NuE2xXm2MsQQkBVF/Halo-UI-1.0.0?type=design&node-id=876-14110&mode=design&t=zYpAhBY2yjbA57Sl-0',
    },
  },
};
