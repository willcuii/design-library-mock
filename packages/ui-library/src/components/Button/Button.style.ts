import {styled} from '@mui/material/styles';
import Button from '@mui/material/Button';
import {ButtonProps} from './Button.type';
import {backgroundColors} from '../../tokens/colors/colorsOnWhite';
import {gradientColors} from '../../tokens/colors/colorPalette';

const StyledButton: React.ComponentType<ButtonProps> = styled(
  Button,
)<ButtonProps>(({theme, btnVariant, btnSize, disabled}) => ({
  // Common Style
  borderRadius: '50px',
  boxShadow: 'none',
  textTransform: 'none',
  fontSize: 16,
  fontWeight: 600,
  lineHeight: 1,
  fontFamily: 'sofia-pro, sans-serif',
  '&:hover, &:active, &:focus': {
    boxShadow: 'none',
  },
  // Primay Style
  ...(btnVariant === 'primary' && {
    backgroundColor: theme.palette.primary.main,
    borderColor: theme.palette.grey[300],
    color: theme.palette.primary.contrastText,
    '&:hover': {
      backgroundColor: theme.palette.primary.light,
      borderColor: theme.palette.primary.light,
    },
    '&:active': {
      backgroundColor: theme.palette.primary.dark,
      borderColor: theme.palette.primary.dark,
    },
    '&:disabled': {
      backgroundColor: backgroundColors.bgButtonDisabled,
      borderColor: backgroundColors.bgButtonDisabled,
      color: theme.palette.text.disabled,
      cursor: 'not-allowed',
      pointerEvents: 'auto',
    },
  }),
  // Outlined Style
  ...(btnVariant === 'outlined' && {
    border: '1px solid',
    backgroundColor: theme.palette.common.white,
    borderColor: theme.palette.grey[300],
    color: theme.palette.text.primary,
    '&:hover': {
      backgroundColor: theme.palette.secondary.light,
      borderColor: theme.palette.secondary.light,
      color: theme.palette.primary.dark,
    },
    '&:active': {
      // TODO overlay?
      backgroundColor: theme.palette.secondary.light,
      borderColor: theme.palette.secondary.light,
      color: theme.palette.primary.dark,
    },
    '&:disabled': {
      backgroundColor: theme.palette.common.white,
      borderColor: theme.palette.grey[100],
      color: theme.palette.text.disabled,
      cursor: 'not-allowed',
      pointerEvents: 'auto',
    },
  }),
  // Text Style
  ...(btnVariant === 'text' && {
    backgroundColor: theme.palette.common.white,
    borderColor: theme.palette.common.white,
    color: theme.palette.text.primary,
    '&:hover': {
      backgroundColor: theme.palette.grey[50],
      borderColor: theme.palette.grey[50],
    },
    '&:active': {
      backgroundColor: theme.palette.action.active,
      borderColor: theme.palette.action.active,
    },
    '&:disabled': {
      backgroundColor: theme.palette.common.white,
      borderColor: theme.palette.common.white,
      color: theme.palette.text.disabled,
      cursor: 'not-allowed',
      pointerEvents: 'auto',
    },
  }),
  // Gradient Style
  ...(btnVariant === 'gradient' && {
    position: 'relative',
    border: 'none',
    color: theme.palette.text.primary,
    background: theme.palette.common.white,
    ...(!disabled && {
      '&::before': {
        content: "''",
        position: 'absolute',
        inset: '0',
        borderRadius: '50px',
        padding: '2px' /* control the border thickness */,
        background: gradientColors.Summer2,
        '-webkit-mask':
          'linear-gradient(#fff 0 0) content-box, linear-gradient(#fff 0 0)',
        '-webkit-mask-composite': 'xor',
        'mask-composite': 'exclude',
        'pointer-events': 'none',
      },
      '&:hover': {
        background: theme.palette.common.white,
        '.gradient-bg::before': {
          content: "''",
          zIndex: -1,
          position: 'absolute',
          top: 0,
          right: 0,
          bottom: 0,
          left: 0,
          borderRadius: '50px',
          background: gradientColors.Summer2,
          transform: 'translate3d(0px, 0px, 0) scale(0.97)',
          filter: 'blur(10px)',
          opacity: 0.7,
          transition: 'opacity 0.3s',
        },
        '.gradient-bg::after': {
          content: "''",
          zIndex: -1,
          position: 'absolute',
          top: 0,
          right: 0,
          bottom: 0,
          left: 0,
          background: 'inherit',
          borderRadius: 'inherit',
        },
      },
      '&:active': {
        color: theme.palette.common.white,
        background: gradientColors.Summer2,
      },
    }),
    '&:disabled': {
      backgroundColor: backgroundColors.bgButtonDisabled,
      borderColor: backgroundColors.bgButtonDisabled,
      color: theme.palette.text.disabled,
      cursor: 'not-allowed',
      pointerEvents: 'auto',
    },
  }),
  // Size Style
  ...(btnSize === 'xsmall' && {
    height: 32,
    padding: '6px 12px',
    fontSize: 14,
  }),
  ...(btnSize === 'small' && {
    height: 40,
    padding: '6px 16px',
    fontSize: 14,
  }),
  ...(btnSize === 'medium' && {
    height: 48,
    padding: '8px 24px',
    fontSize: 16,
  }),
  ...(btnSize === 'large' && {
    height: 56,
    padding: '12px 36px',
    fontSize: 16,
  }),
}));

export default StyledButton;
