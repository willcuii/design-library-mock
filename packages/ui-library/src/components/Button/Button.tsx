import StyledButton from './Button.style';
import {ButtonProps} from './Button.type';

const Button = ({children, onClick, id}: ButtonProps) => {
  return (
    <StyledButton onClick={onClick} id={id}>
      {children}
    </StyledButton>
  );
};

Button.displayName = 'Button';
export default Button;
