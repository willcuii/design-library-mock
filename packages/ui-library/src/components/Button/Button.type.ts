import {ReactNode} from 'react';
import {ButtonProps as MuiButtonProps} from '@mui/material/Button';

export type ButtonSize = 'xsmall' | 'small' | 'medium' | 'large';

export type ButtonVariant = 'primary' | 'outlined' | 'text' | 'gradient';

export type LoadingButtonPostion = 'start' | 'center' | 'end';

export interface ButtonProps extends MuiButtonProps {
  /** The content of the component. */
  children: ReactNode;

  /**
   * A button can show that it cannot be interacted with.
   *
   * @default false
   * */
  disabled?: boolean;

  /** A button can have an icon as the 'children' element */
  icon?: ReactNode;

  /**
   * A button can be sized to fit different layout.
   *
   * @default 'medium'
   * */
  btnSize?: ButtonSize;

  /**
   * The `btnVariant` prop defines the visual styling of a button, allowing it to stand out or blend in according to its intended role within the user interface.
   * Each variant modifies the button's appearance to suit different contexts and actions.
   *
   * @default 'primary'
   */
  btnVariant?: 'primary' | 'outlined' | 'text' | 'gradient';

  /** Override or extend the styles applied to the component. */
  classes?: MuiButtonProps['classes'];

  /** Icon element placed before the children. */
  startIcon?: ReactNode;

  /** Icon element placed after the children. */
  endIcon?: ReactNode;

  /** Function to be trigger when user click button */
  onClick?: MuiButtonProps['onClick'];
}

export interface LoadingButtonProps extends ButtonProps {
  /**
   * The loading indicator to use.
   *
   * @default '<CircularProgress size={24} />'
   */
  loadingIndicator?: ReactNode;

  /**
   * If true, the loading indicator is shown and the button becomes disabled.
   *
   * @default true
   * */
  loading?: boolean;

  /**
   * The loading indicator can be positioned on the start, end, or the center of the button.
   *
   * @default 'center'
   */
  loadingPosition?: LoadingButtonPostion;
}
