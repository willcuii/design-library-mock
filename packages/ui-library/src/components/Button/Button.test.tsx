import React from 'react';
import {describe, expect, it, vi} from 'vitest';
import userEvent from '@testing-library/user-event';
import {render, fireEvent} from '@testing-library/react';
import Button from './Button';

describe('Button Unit Tests', () => {
  describe('when rendered as a button', () => {
    it('renders the button with the provided text', () => {
      const {getByText} = render(<Button>Click me</Button>);
      expect(getByText('Click me')).toBeDefined();
    });

    it('can be focused', () => {
      const {getByRole} = render(<Button>This is a button</Button>);
      const button = getByRole('button');

      expect(document.activeElement).not.toEqual(button);
      userEvent.tab();
      expect(document.activeElement).toEqual(button);
    });

    it('cannot be focused when disabled has been passed to the component', () => {
      const {getByRole} = render(<Button disabled>This is a button</Button>);
      const button = getByRole('button');

      expect(document.activeElement).not.toEqual(button);
      userEvent.tab();
      expect(document.activeElement).not.toEqual(button);
    });

    it('can trigger the onClick function when clicked', () => {
      const onClick = vi.fn();
      const {getByText} = render(<Button onClick={onClick}>Click me</Button>);
      fireEvent.click(getByText('Click me'));
      expect(onClick).toHaveBeenCalled();
    });

    it('can not trigger the onClick function by being clicked when disabled', () => {
      const onClick = vi.fn();
      const {getByText} = render(
        <Button disabled onClick={onClick}>
          Click me
        </Button>,
      );

      fireEvent.click(getByText('Click me'));
      expect(onClick).not.toHaveBeenCalled();
    });

    it('renders the button as disabled when disabled prop is true', () => {
      const {getByText} = render(<Button disabled>Click me</Button>);
      expect(getByText('Click me')).toHaveProperty('disabled', true);
    });
  });

  // it('renders the button with the primary btnVariant', () => {
  //   const {getByText} = render(<Button btnVariant="primary">Click me</Button>);
  //   expect(getByText('Click me')).toMatchSnapshot();
  // });

  // it('renders the button with the outlined btnVariant', () => {
  //   const {getByText} = render(<Button btnVariant="primary">Click me</Button>);
  //   expect(getByText('Click me')).toMatchSnapshot();
  // });

  // it('renders the button with the provided btnSize', () => {
  //   const {getByText} = render(<Button btnVariant="outlined">Click me</Button>);
  //   expect(getByText('Click me')).toMatchSnapshot();
  // });
});
