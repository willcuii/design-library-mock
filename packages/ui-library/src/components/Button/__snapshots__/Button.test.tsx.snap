// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Unit Tests > renders the button with the outlined btnVariant 1`] = `
<button
  btnsize="medium"
  btnvariant="primary"
  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1ieilu9-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  <div
    class="gradient-bg"
  />
  Click me
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`Button Unit Tests > renders the button with the primary btnVariant 1`] = `
<button
  btnsize="medium"
  btnvariant="primary"
  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1ieilu9-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  <div
    class="gradient-bg"
  />
  Click me
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`Button Unit Tests > renders the button with the provided btnSize 1`] = `
<button
  btnsize="medium"
  btnvariant="outlined"
  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1evtol0-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  <div
    class="gradient-bg"
  />
  Click me
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`Button Unit Tests > renders the button with the provided btnVariant 1`] = `
<button
  btnsize="medium"
  btnvariant="primary"
  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1ieilu9-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  <div
    class="gradient-bg"
  />
  Click me
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;
